cmds/tr/tr.go

```diff
@@ -0,0 +1,306 @@
+	'r':  '\r',
+	't':  '\t',
+	'v':  '\v',
+}
+
+type Set int
+
+const (
+	ALNUM Set = iota
+	ALPHA
+	DIGIT
+	GRAPH
+	CNTRL
+	PUNCT
+	LOWER
+	UPPER
+	SPACE
+
+	UNKNOWN
+)
+
+func cmdToSet(cmd string) Set {
+	switch cmd {
+	case "[:alnum:]":
+		return ALNUM
+	case "[:alpha:]":
+		return ALPHA
+	case "[:digit:]":
+		return DIGIT
+	case "[:graph:]":
+		return GRAPH
+	case "[:cntrl:]":
+		return CNTRL
+	case "[:punct:]":
+		return PUNCT
+	case "[:lower:]":
+		return LOWER
+	case "[:upper:]":
+		return UPPER
+	case "[:space:]":
+		return SPACE
+	default:
+		return UNKNOWN
+	}
+
+}
+
+type transformer struct {
+	transform func(r rune) rune
+}
+
+func setToRune(s Set, outRune rune) transformer {
+	var check func(r rune) bool
+	switch s {
+	case ALNUM:
+		check = func(r rune) bool {
+			return unicode.IsLetter(r) || unicode.IsDigit(r)
+		}
+	case ALPHA:
+		check = unicode.IsLetter
+	case DIGIT:
+		check = unicode.IsDigit
+	case GRAPH:
+		check = unicode.IsGraphic
+	case CNTRL:
+		check = unicode.IsControl
+	case PUNCT:
+		check = unicode.IsPunct
+	case LOWER:
+		check = unicode.IsLower
+	case UPPER:
+		check = unicode.IsUpper
+	case SPACE:
+		check = unicode.IsSpace
+	}
+	return transformer{
+		transform: func(r rune) rune {
+			if check(r) {
+				return outRune
+			}
+			return r
+		},
+	}
+}
+
+func lowerToUpper() transformer {
+	return transformer{
+		transform: func(r rune) rune {
+			return unicode.ToUpper(r)
+		},
+	}
+}
+
+func upperToLower() transformer {
+	return transformer{
+		transform: func(r rune) rune {
+			return unicode.ToLower(r)
+		},
+	}
+}
+
+func runesToRunes(in []rune, out ...rune) transformer {
+	convs := make(map[rune]rune)
+	l := len(out)
+	for i, r := range in {
+		ind := i
+		if i > l-1 {
+			ind = l - 1
+		}
+		convs[r] = out[ind]
+	}
+	return transformer{
+		transform: func(r rune) rune {
+			if outRune, ok := convs[r]; ok {
+				return outRune
+			}
+			return r
+		},
+	}
+}
+
+func (t transformer) run(r io.Reader, w io.Writer) error {
+	in := bufio.NewReader(r)
+	out := bufio.NewWriter(w)
+
+	defer out.Flush()
+
+	for {
+		inRune, size, er := in.ReadRune()
+		if inRune == unicode.ReplacementChar {
+			// can skip error hanling here, because
+			// previous operation was in.ReadRune()
+			in.UnreadRune()
+
+			b, err := in.ReadByte()
+			if err != nil {
+				return fmt.Errorf("read error: %v", err)
+			}
+
+			if err := out.WriteByte(b); err != nil {
+				return fmt.Errorf("write error: %v", err)
+			}
+		} else if size > 0 {
+			if outRune := t.transform(inRune); outRune != unicode.ReplacementChar {
+				if _, err := out.WriteRune(outRune); err != nil {
+					return fmt.Errorf("write error: %v", err)
+				}
+			}
+		}
+
+		if er != nil {
+			if er == io.EOF {
+				return nil
+			}
+			return er
+		}
+	}
+}
+
+func parse() (transformer, error) {
+	flag.Parse()
+
+	narg := flag.NArg()
+	args := flag.Args()
+	switch {
+	case narg == 0 || (narg == 1 && !*delete):
+		return transformer{}, fmt.Errorf("missing operand")
```

Comment:
then all instances of transformer{} can be nil

Reply:
done